lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+    hook_fields[kDestroy] -= +!!this[destroy_symbol];
+    hooks_array.splice(index, 1);
+    return this;
+  }
+}
+
+
+function storeActiveHooks() {
+  tmp_active_hooks_array = active_hooks_array.slice();
+  // Don't want to make the assumption that kInit to kDestroy are indexes 0 to
+  // 4. So do this the long way.
+  tmp_async_hook_fields = [];
+  tmp_async_hook_fields[kInit] = async_hook_fields[kInit];
+  tmp_async_hook_fields[kBefore] = async_hook_fields[kBefore];
+  tmp_async_hook_fields[kAfter] = async_hook_fields[kAfter];
+  tmp_async_hook_fields[kDestroy] = async_hook_fields[kDestroy];
+}
+
+
+// Then restore the correct hooks array in case any hooks were added/removed
+// during hook callback execution.
+function restoreTmpHooks() {
+  active_hooks_array = tmp_active_hooks_array;
+  async_hook_fields[kInit] = tmp_async_hook_fields[kInit];
+  async_hook_fields[kBefore] = tmp_async_hook_fields[kBefore];
+  async_hook_fields[kAfter] = tmp_async_hook_fields[kAfter];
+  async_hook_fields[kDestroy] = tmp_async_hook_fields[kDestroy];
+
+  tmp_active_hooks_array = null;
+  tmp_async_hook_fields = null;
+}
+
+
+function createHook(fns) {
+  return new AsyncHook(fns);
+}
+
+
+function currentId() {
+  return async_uid_fields[kCurrentAsyncId];
+}
+
+
+function triggerId() {
+  return async_uid_fields[kCurrentTriggerId];
+}
+
+
+// Embedder API //
+
+class AsyncEvent {
+  constructor(type, triggerId) {
+    this[async_id_symbol] = ++async_uid_fields[kAsyncUidCntr];
+    // Do this before early return to null any potential kInitTriggerId.
+    if (triggerId === undefined)
+      triggerId = initTriggerId();
+    // If a triggerId was passed, any kInitTriggerId still must be null'd.
+    else
+      async_uid_fields[kInitTriggerId] = 0;
+    this[trigger_id_symbol] = triggerId;
+
+    // Return immediately if there's nothing to do.
+    if (async_hook_fields[kInit] === 0)
+      return;
+
+    if (typeof type !== 'string' || type.length <= 0)
+      throw new TypeError('type must be a string with length > 0');
+    if (!Number.isSafeInteger(triggerId) || triggerId < 0)
+      throw new RangeError('triggerId must be an unsigned integer');
+
+    for (var i = 0; i < active_hooks_array.length; i++) {
+      if (typeof active_hooks_array[i][init_symbol] === 'function') {
+        runInitCallback(active_hooks_array[i][init_symbol],
+                        this[async_id_symbol],
+                        type,
+                        triggerId,
+                        this);
+      }
+    }
+  }
+
+  emitBefore() {
+    emitBeforeS(this[async_id_symbol], this[trigger_id_symbol]);
+    return this;
+  }
+
+  emitAfter() {
+    emitAfterS(this[async_id_symbol]);
+    return this;
+  }
+
+  emitDestroy() {
+    emitDestroyS(this[async_id_symbol]);
+    return this;
+  }
+
+  asyncId() {
+    return this[async_id_symbol];
+  }
+
+  triggerId() {
+    return this[trigger_id_symbol];
+  }
+}
+
+
+function asyncIdScope(asyncId, cb) {
```

Comment:
I'm fine with the name change. As far as the restful args, I'll benchmark. Haven't used them lately so not sure how they'll hold up.